File: sdk-js/src/accumulator.test.ts
import test, { suite, type TestContext } from "node:test";
import { StreamAccumulator } from "./accumulator.ts";
import { InvariantError } from "./errors.ts";

suite("StreamAccumulator", () => {
  test("reconstructs interleaved multipart streams and metadata", (t: TestContext) => {
    const accumulator = new StreamAccumulator();

    accumulator.addPartial({
      delta: { index: 2, part: { type: "reasoning", text: "think " } },
      usage: { input_tokens: 2, output_tokens: 1 },
      cost: 0.1,
    });
    accumulator.addPartial({
      delta: { index: 0, part: { type: "text", text: "Hel" } },
    });
    accumulator.addPartial({
      delta: {
        index: 1,
        part: {
          type: "tool-call",
          tool_call_id: "call_1",
          call: { type: "function", name: "weather", args: '{"city":' },
        },
      },
    });
    accumulator.addPartial({
      delta: { index: 0, part: { type: "text", text: "lo" } },
      usage: { input_tokens: 3, output_tokens: 4 },
      cost: 0.2,
    });
    accumulator.addPartial({
      delta: {
        index: 1,
        part: {
          type: "tool-call",
          call: { type: "function", args: '"Paris"}' },
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 2,
        part: { type: "reasoning", text: "done", signature: "sig" },
      },
    });

    t.assert.strictEqual(accumulator.size, 3);
    t.assert.strictEqual(accumulator.isEmpty, false);
    t.assert.deepStrictEqual(accumulator.computeResponse(), {
      content: [
        { type: "text", text: "Hello" },
        {
          type: "tool-call",
          tool_call_id: "call_1",
          call: { type: "function", name: "weather", args: { city: "Paris" } },
        },
        { type: "reasoning", text: "think done", signature: "sig" },
      ],
      usage: { input_tokens: 5, output_tokens: 5 },
      cost: 0.30000000000000004,
    });
  });

  test("rejects a different part type at an occupied index", (t: TestContext) => {
    const accumulator = new StreamAccumulator();
    accumulator.addPartial({
      delta: { index: 0, part: { type: "text", text: "hello" } },
    });

    t.assert.throws(() => {
      accumulator.addPartial({
        delta: {
          index: 0,
          part: { type: "reasoning", text: "wrong type" },
        },
      });
    }, /Type mismatch at index 0/);
  });

  test("rejects malformed and incomplete tool calls", (t: TestContext) => {
    const malformed = new StreamAccumulator();
    malformed.addPartial({
      delta: {
        index: 0,
        part: {
          type: "tool-call",
          tool_call_id: "call_1",
          call: { type: "function", name: "weather", args: "{bad json" },
        },
      },
    });
    t.assert.throws(
      () => malformed.computeResponse(),
      (error: unknown) => error instanceof InvariantError,
    );

    const incomplete = new StreamAccumulator();
    incomplete.addPartial({
      delta: {
        index: 0,
        part: { type: "tool-call", call: { type: "function", args: "{}" } },
      },
    });
    t.assert.throws(
      () => incomplete.computeResponse(),
      /Missing required fields/,
    );
  });

  test("snapshots independently materializable parts", (t: TestContext) => {
    const accumulator = new StreamAccumulator();
    accumulator.addPartial({
      delta: { index: 0, part: { type: "text", text: "partial" } },
      usage: { input_tokens: 2, output_tokens: 3 },
      cost: 0.25,
    });
    accumulator.addPartial({
      delta: {
        index: 1,
        part: {
          type: "tool-call",
          tool_call_id: "call_1",
          call: {
            type: "function",
            name: "weather",
            args: '{"city":"Paris"}',
          },
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 2,
        part: {
          type: "tool-call",
          call: { type: "function", args: "{incomplete" },
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 3,
        part: {
          type: "image",
          data: "aGVsbG8=",
          mime_type: "image/png",
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 4,
        part: { type: "audio", data: "AAABAA==", format: "linear16" },
      },
    });

    t.assert.deepStrictEqual(accumulator.snapshot(), {
      content: [
        { type: "text", text: "partial" },
        {
          type: "tool-call",
          tool_call_id: "call_1",
          call: {
            type: "function",
            name: "weather",
            args: { city: "Paris" },
          },
        },
        { type: "image", data: "aGVsbG8=", mime_type: "image/png" },
        { type: "audio", data: "AAABAA==", format: "linear16" },
      ],
      usage: { input_tokens: 2, output_tokens: 3 },
      cost: 0.25,
    });
    t.assert.throws(() => accumulator.computeResponse());
  });

  test("accumulates hosted web-search status and atomic results", (t: TestContext) => {
    const accumulator = new StreamAccumulator();
    accumulator.addPartial({
      delta: {
        index: 0,
        part: {
          type: "tool-call",
          tool_call_id: "ws_1",
          call: { type: "web_search", status: "in_progress" },
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 0,
        part: {
          type: "tool-call",
          call: {
            type: "web_search",
            status: "completed",
            action: { type: "search", queries: ["sdk docs"] },
          },
        },
      },
    });
    accumulator.addPartial({
      delta: {
        index: 1,
        part: {
          type: "tool-result",
          tool_call_id: "ws_1",
          result: {
            type: "web_search",
            sources: [
              {
                url: "https://example.com",
                title: "Example",
                signature: "opaque",
              },
            ],
          },
          status: "completed",
        },
      },
    });
    t.assert.deepStrictEqual(accumulator.computeResponse().content, [
      {
        type: "tool-call",
        tool_call_id: "ws_1",
        call: {
          type: "web_search",
          status: "completed",
          action: { type: "search", queries: ["sdk docs"] },
        },
      },
      {
        type: "tool-result",
        tool_call_id: "ws_1",
        result: {
          type: "web_search",
          sources: [
            {
              url: "https://example.com",
              title: "Example",
              signature: "opaque",
            },
          ],
        },
        status: "completed",
      },
    ]);
  });

  test("preserves hosted web-search error codes", (t: TestContext) => {
    const accumulator = new StreamAccumulator();
    accumulator.addPartial({
      delta: {
        index: 0,
        part: {
          type: "tool-result",
          tool_call_id: "ws_error",
          result: {
            type: "web_search",
            sources: [],
            error_code: "unavailable",
          },
          status: "failed",
        },
      },
    });
    t.assert.deepStrictEqual(accumulator.computeResponse().content, [
      {
        type: "tool-result",
        tool_call_id: "ws_error",
        result: {
          type: "web_search",
          sources: [],
          error_code: "unavailable",
        },
        status: "failed",
      },
    ]);
  });
});

File: sdk-go/accumulator_test.go
package llmsdk_test

import (
	"testing"

	"github.com/google/go-cmp/cmp"
	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
	"github.com/hoangvvo/llm-sdk/sdk-go/utils/ptr"
)

func TestStreamAccumulator_ReconstructsInterleavedMultipartStream(t *testing.T) {
	accumulator := llmsdk.NewStreamAccumulator()
	partials := []llmsdk.PartialModelResponse{
		{
			Delta: &llmsdk.ContentDelta{Index: 2, Part: llmsdk.NewReasoningPartDelta("think ")},
			Usage: &llmsdk.ModelUsage{InputTokens: 2, OutputTokens: 1},
			Cost:  ptr.To(0.1),
		},
		{Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("Hel")}},
		{Delta: &llmsdk.ContentDelta{Index: 1, Part: llmsdk.NewToolCallPartDelta(
			llmsdk.WithToolCallPartDeltaToolCallID("call_1"),
			llmsdk.WithToolCallPartDeltaToolName("weather"),
			llmsdk.WithToolCallPartDeltaArgs(`{"city":`),
		)}},
		{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("lo")},
			Usage: &llmsdk.ModelUsage{InputTokens: 3, OutputTokens: 4},
			Cost:  ptr.To(0.2),
		},
		{Delta: &llmsdk.ContentDelta{Index: 1, Part: llmsdk.NewToolCallPartDelta(
			llmsdk.WithToolCallPartDeltaArgs(`"Paris"}`),
		)}},
		{Delta: &llmsdk.ContentDelta{Index: 2, Part: llmsdk.NewReasoningPartDelta(
			"done",
			llmsdk.WithReasoningPartDeltaSignature("sig"),
		)}},
	}

	for _, partial := range partials {
		if err := accumulator.AddPartial(partial); err != nil {
			t.Fatalf("add partial: %v", err)
		}
	}
	if accumulator.Size() != 3 || accumulator.IsEmpty() {
		t.Fatalf("unexpected accumulator state: size=%d empty=%v", accumulator.Size(), accumulator.IsEmpty())
	}

	response, err := accumulator.ComputeResponse()
	if err != nil {
		t.Fatalf("compute response: %v", err)
	}
	expected := llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			llmsdk.NewTextPart("Hello"),
			llmsdk.NewToolCallPart("call_1", "weather", map[string]any{"city": "Paris"}),
			llmsdk.NewReasoningPart("think done", llmsdk.WithReasoningSignature("sig")),
		},
		Usage: &llmsdk.ModelUsage{InputTokens: 5, OutputTokens: 5},
		Cost:  ptr.To(0.30000000000000004),
	}
	if diff := cmp.Diff(expected, response); diff != "" {
		t.Fatalf("response mismatch (-want +got):\n%s", diff)
	}
}

func TestStreamAccumulator_RejectsMismatchedAndMalformedDeltas(t *testing.T) {
	t.Run("part type changes", func(t *testing.T) {
		accumulator := llmsdk.NewStreamAccumulator()
		if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("hello")},
		}); err != nil {
			t.Fatalf("add initial partial: %v", err)
		}
		if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewReasoningPartDelta("wrong")},
		}); err == nil {
			t.Fatal("expected a part-type mismatch error")
		}
	})

	t.Run("tool call kind changes", func(t *testing.T) {
		functionDelta := func() llmsdk.PartDelta {
			return llmsdk.NewToolCallPartDelta(
				llmsdk.WithToolCallPartDeltaToolCallID("call_1"),
				llmsdk.WithToolCallPartDeltaToolName("lookup"),
				llmsdk.WithToolCallPartDeltaArgs("{}"),
			)
		}
		webSearchDelta := func() llmsdk.PartDelta {
			id := "call_1"
			return llmsdk.PartDelta{ToolCallPartDelta: &llmsdk.ToolCallPartDelta{
				ToolCallID: &id,
				Call: llmsdk.ToolCallDelta{WebSearch: &llmsdk.WebSearchToolCallDelta{
					Action: &llmsdk.WebSearchAction{Type: "search", Queries: []string{"sdk docs"}},
				}},
			}}
		}

		for _, deltas := range [][]llmsdk.PartDelta{
			{functionDelta(), webSearchDelta()},
			{webSearchDelta(), functionDelta()},
		} {
			accumulator := llmsdk.NewStreamAccumulator()
			if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
				Delta: &llmsdk.ContentDelta{Index: 0, Part: deltas[0]},
			}); err != nil {
				t.Fatalf("add initial partial: %v", err)
			}
			if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
				Delta: &llmsdk.ContentDelta{Index: 0, Part: deltas[1]},
			}); err == nil {
				t.Fatal("expected a tool-call kind mismatch error")
			}
		}
	})

	t.Run("malformed tool arguments", func(t *testing.T) {
		accumulator := llmsdk.NewStreamAccumulator()
		if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewToolCallPartDelta(
				llmsdk.WithToolCallPartDeltaToolCallID("call_1"),
				llmsdk.WithToolCallPartDeltaToolName("weather"),
				llmsdk.WithToolCallPartDeltaArgs("{bad json"),
			)},
		}); err != nil {
			t.Fatalf("add partial: %v", err)
		}
		if _, err := accumulator.ComputeResponse(); err == nil {
			t.Fatal("expected malformed tool arguments to fail")
		}
	})

	t.Run("large numeric tool arguments", func(t *testing.T) {
		accumulator := llmsdk.NewStreamAccumulator()
		if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewToolCallPartDelta(
				llmsdk.WithToolCallPartDeltaToolCallID("call_1"),
				llmsdk.WithToolCallPartDeltaToolName("lookup"),
				llmsdk.WithToolCallPartDeltaArgs(`{"id":9007199254740993}`),
			)},
		}); err != nil {
			t.Fatal(err)
		}
		response, err := accumulator.ComputeResponse()
		if err != nil {
			t.Fatal(err)
		}
		args := response.Content[0].ToolCallPart.Call.Function.Args
		if string(args) != `{"id":9007199254740993}` {
			t.Fatalf("tool arguments changed while accumulating: %s", args)
		}
	})
}

func TestStreamAccumulator_ClearRemovesContentAndMetadata(t *testing.T) {
	accumulator := llmsdk.NewStreamAccumulator()
	if err := accumulator.AddPartial(llmsdk.PartialModelResponse{
		Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("old")},
		Usage: &llmsdk.ModelUsage{InputTokens: 2, OutputTokens: 1},
		Cost:  ptr.To(0.4),
	}); err != nil {
		t.Fatalf("add partial: %v", err)
	}

	accumulator.Clear()
	response, err := accumulator.ComputeResponse()
	if err != nil {
		t.Fatalf("compute cleared response: %v", err)
	}
	if !accumulator.IsEmpty() || response.Usage != nil || response.Cost != nil || len(response.Content) != 0 {
		t.Fatalf("clear retained state: %#v", response)
	}
}

func TestStreamAccumulator_SnapshotsIndependentlyMaterializableParts(t *testing.T) {
	accumulator := llmsdk.NewStreamAccumulator()
	partials := []llmsdk.PartialModelResponse{
		{
			Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("partial")},
			Usage: &llmsdk.ModelUsage{InputTokens: 2, OutputTokens: 3},
			Cost:  ptr.To(0.25),
		},
		{Delta: &llmsdk.ContentDelta{Index: 1, Part: llmsdk.NewToolCallPartDelta(
			llmsdk.WithToolCallPartDeltaToolCallID("call_1"),
			llmsdk.WithToolCallPartDeltaToolName("weather"),
			llmsdk.WithToolCallPartDeltaArgs(`{"city":"Paris"}`),
		)}},
		{Delta: &llmsdk.ContentDelta{Index: 2, Part: llmsdk.NewToolCallPartDelta(
			llmsdk.WithToolCallPartDeltaArgs("{incomplete"),
		)}},
		{Delta: &llmsdk.ContentDelta{Index: 3, Part: llmsdk.NewImagePartDelta(
			llmsdk.WithImagePartDeltaData("aGVsbG8="),
			llmsdk.WithImagePartDeltaMimeType("image/png"),
		)}},
		{Delta: &llmsdk.ContentDelta{Index: 4, Part: llmsdk.NewAudioPartDelta(
			llmsdk.WithAudioPartDeltaData("AAABAA=="),
			llmsdk.WithAudioPartDeltaFormat(llmsdk.AudioFormatLinear16),
		)}},
	}
	for _, partial := range partials {
		if err := accumulator.AddPartial(partial); err != nil {
			t.Fatalf("add partial: %v", err)
		}
	}

	expected := llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			llmsdk.NewTextPart("partial"),
			llmsdk.NewToolCallPart("call_1", "weather", map[string]any{"city": "Paris"}),
			llmsdk.NewImagePart("aGVsbG8=", "image/png"),
			llmsdk.NewAudioPart("AAABAA==", llmsdk.AudioFormatLinear16),
		},
		Usage: &llmsdk.ModelUsage{InputTokens: 2, OutputTokens: 3},
		Cost:  ptr.To(0.25),
	}
	if diff := cmp.Diff(expected, accumulator.Snapshot()); diff != "" {
		t.Fatalf("snapshot mismatch (-want +got):\n%s", diff)
	}
	if _, err := accumulator.ComputeResponse(); err == nil {
		t.Fatal("strict response unexpectedly accepted an incomplete part")
	}
}

func TestStreamAccumulator_WebSearchCallAndResult(t *testing.T) {
	accumulator := llmsdk.NewStreamAccumulator()
	status := llmsdk.WebSearchToolCallStatusCompleted
	id := "ws_1"
	partials := []llmsdk.PartialModelResponse{
		{Delta: &llmsdk.ContentDelta{
			Index: 0,
			Part: llmsdk.PartDelta{ToolCallPartDelta: &llmsdk.ToolCallPartDelta{
				ToolCallID: &id,
				Call: llmsdk.ToolCallDelta{WebSearch: &llmsdk.WebSearchToolCallDelta{
					Status: &status,
					Action: &llmsdk.WebSearchAction{Type: "search", Queries: []string{"sdk docs"}},
				}},
			}},
		}},
		{Delta: &llmsdk.ContentDelta{
			Index: 1,
			Part: llmsdk.PartDelta{ToolResultPartDelta: &llmsdk.ToolResultPartDelta{
				ToolCallID: id,
				Result: llmsdk.ToolResult{WebSearch: &llmsdk.WebSearchToolResult{
					Sources: []llmsdk.WebSearchSource{{URL: "https://example.com"}},
				}},
				Status: llmsdk.ToolResultStatusCompleted,
			}},
		}},
	}
	for _, partial := range partials {
		if err := accumulator.AddPartial(partial); err != nil {
			t.Fatal(err)
		}
	}
	response, err := accumulator.ComputeResponse()
	if err != nil {
		t.Fatal(err)
	}
	if len(response.Content) != 2 || response.Content[0].ToolCallPart.Call.WebSearch == nil || response.Content[1].ToolResultPart.Result.WebSearch == nil {
		t.Fatalf("unexpected response: %#v", response.Content)
	}
}
